pallets/free-calls/src/lib.rs

```diff
@@ -0,0 +1,430 @@
+                    Some(non_zero) => non_zero,
+                    None => panic!("quota_ratio must be non zero"),
+                },
+            }
+        }
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub (super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_locker_mirror::Config {
+        /// The overarching event type.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// The call type from the runtime which has all the calls available in your runtime.
+        type Call: Parameter
+            + Dispatchable<Origin = Self::Origin, PostInfo = PostDispatchInfo>
+            + GetDispatchInfo
+            + From<frame_system::Call<Self>>
+            + IsSubType<Call<Self>>
+            + IsType<<Self as frame_system::Config>::Call>;
+
+        /// The configurations that will be used to limit the usage of the allocated quota to these
+        /// different configs.
+        #[pallet::constant]
+        type WindowsConfig: Get<Vec<WindowConfig<Self::BlockNumber>>>;
+
+        /// Filter on which calls are permitted to be free.
+        type CallFilter: Contains<<Self as Config>::Call>;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+
+        /// A calculation strategy to convert locked tokens info to a quota.
+        type QuotaCalculationStrategy: QuotaCalculationStrategy<Self>;
+    }
+
+    /// Keeps track of each windows usage for each consumer.
+    #[pallet::storage]
+    #[pallet::getter(fn window_stats_by_consumer)]
+    pub(super) type WindowStatsByConsumer<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        T::AccountId,
+        Twox64Concat,
+        // Index of the window in the list of window configurations.
+        WindowType,
+        ConsumerStats<T::BlockNumber>,
+    >;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub (super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Free call was executed. [who, result]
+        FreeCallResult(T::AccountId, DispatchResult),
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+
+        /// Try to execute a call using the free allocated quota. This call may not execute because
+        /// one of the following reasons:
+        ///  * Caller has no free quota set.
+        ///  * The caller has used all the allowed quota for at least one window config.
+        #[pallet::weight({
+            let boxed_call_info = call.get_dispatch_info();
+            let boxed_call_weight = boxed_call_info.weight;
+            let self_weight = <T as Config>::WeightInfo::try_free_call();
+
+            let total_weight = self_weight.saturating_add(boxed_call_weight);
+            (
+                total_weight,
+                boxed_call_info.class,
+                Pays::No,
+            )
```

Comment:
please put a note that you have a signed extension doing a pre-validation check here